src/test/java/org/mskcc/cbio/oncokb/web/rest/UsageAnalysisControllerIT.java

```diff
@@ -0,0 +1,931 @@
+      }
+    }
+
+    private final String[] sampleOncokbEndpoints = {
+      "/api/v1/annotate/mutations/byProteinChange",
+      "/api/private/utils/numbers/main/",
+      "/api/v1/annotate/structuralVariants",
+      "/api/private/utils/numbers/levels/",
+      "/api/v1/annotate/copyNumberAlterations",
+      "/api/private/search/typeahead",
+    };
+    private final String[] fakeCompanies = { "FakeCo", "DemoLLC" };
+    private final Random random;
+
+    private final JsonObject files = new JsonObject();
+
+    private final List<User> users = new ArrayList<>();
+    private final List<UserDTO> userDtos = new ArrayList<>();
+
+    public MockS3Data(int seed) throws Exception {
+      this(seed, new String[] { "doej", "smithj" });
+    }
+
+    public MockS3Data(int seed, String[] userNames) throws Exception {
+      random = new Random(seed);
+      addUsers(userNames);
+    }
+
+    private String getNextOncokbEndpoint() {
+      return sampleOncokbEndpoints[Math.abs(
+          random.nextInt() % sampleOncokbEndpoints.length
+        )];
+    }
+
+    /**
+     * Get value from files JSON object based on the JSON path created by the
+     * {@code keys}. If an object is missing in the JSON path then it's created.
+     * @param keys The keys in JSON path
+     * @return The resulting {@code KeyValuePair}
+     */
+    private KeyValuePair<String, JsonObject> fetchJsonObjectFromFilesObject(
+      String... keys
+    ) {
+      JsonObject obj = files;
+      for (int i = 0; i < keys.length - 1; i++) {
+        String key = keys[i];
+        if (!obj.has(key)) {
+          obj.add(key, new JsonObject());
+        }
+        obj = (JsonObject) obj.get(key);
+      }
+      String lastKey = keys[keys.length - 1];
+      return new KeyValuePair<>(lastKey, obj);
+    }
+
+    /**
+     * Get value from files JSON object based on the JSON path created by the
+     * {@code filePath}/[{@code index}]/{@code keys}.
+     * If an object is missing in the JSON path then it's created.
+     * @param filePath First key in the files object.
+     * @param index Array index of the array the {@code filePath} parameter points to.
+     * @param keys The remaining JSON path keys.
+     * @return The resulting {@code KeyValuePair}
+     */
+    private KeyValuePair<String, JsonObject> fetchJsonObjectFromFilesObject(
+      String filePath,
+      int index,
+      String... keys
+    ) {
+      JsonObject obj = files;
+      if (!obj.has(filePath)) {
+        obj.add(filePath, new JsonArray());
+      }
+      JsonArray arr = (JsonArray) obj.get(filePath);
+      while (index >= arr.size()) {
+        arr.add(new JsonObject());
+      }
+      obj = (JsonObject) arr.get(index);
+
+      for (int i = 0; i < keys.length - 1; i++) {
+        String key = keys[i];
+        if (!obj.has(key)) {
+          obj.add(key, new JsonObject());
+        }
+        obj = (JsonObject) obj.get(key);
+      }
+      String lastKey = keys[keys.length - 1];
```

Comment:
Do you intend to let the test fails by throwing out of bound exception when no key been added to this method?

Reply:
It shouldn't happen since this code is just generating the mock data. I don't think people will touch this code very often, but I'll update the methods so that this exception cannot occur.